src/tools/miri/src/machine.rs

```diff
@@ -1225,6 +1226,46 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
         })
     }
 
+    /// Called to adjust global allocations to the Provenance and AllocExtra of this machine.
+    ///
+    /// If `alloc` contains pointers, then they are all pointing to globals.
+    ///
+    /// This should avoid copying if no work has to be done! If this returns an owned
+    /// allocation (because a copy had to be done to adjust things), machine memory will
+    /// cache the result. (This relies on `AllocMap::get_or` being able to add the
+    /// owned allocation to the map even when the map is shared.)
+    fn adjust_global_allocation<'b>(
+        ecx: &InterpCx<'tcx, Self>,
+        id: AllocId,
+        alloc: &'b Allocation,
+    ) -> InterpResult<'tcx, Cow<'b, Allocation<Self::Provenance, Self::AllocExtra, Self::Bytes>>>
+    {
+        // The default implementation does a copy; CTFE machines have a more efficient implementation
+        // based on their particular choice for `Provenance`, `AllocExtra`, and `Bytes`.
+        let kind = Self::GLOBAL_KIND
+            .expect("if GLOBAL_KIND is None, adjust_global_allocation must be overwritten");
+        // This closure is special in that it relies on `take_prepared_alloc_bytes` to give it some
+        // reserved memory space in case Miri needs to handle addresses especially for native calls.
+        let alloc = alloc.adjust_from_tcx(&ecx.tcx, |bytes, align| {
+                Ok(if ecx.machine.native_lib.is_some() {
```

Comment:
Please move this `if` inside `take_prepared_alloc_bytes`, so that the two checks for `native_lib.is_some()` exist both in the same module.

To keep things efficient, it is probably best to also pass `bytes` and `align` into `take_prepared_alloc_bytes` and do everything there (including the sanity-check and copying the bytes into the prepared allocation).

Reply:
Per this suggestion a lot of code ends up moving into the other function;

If we also let `take_prepared_alloc_bytes` (->"`get_alloc_bytes`"?) encapsulate the *`MiriAllocBytes::from_bytes(std::borrow::Cow::Borrowed(&*bytes), align)`* cloning case for us, then this call site could become a lot more simple in general:
```rust
let alloc = alloc.adjust_from_tcx(&ecx.tcx,
    |bytes, align| ecx.take_prepared_alloc_bytes(id, kind, bytes, align),
    |ptr| ecx.global_root_pointer(ptr),
)?;
```

An alternative would be for `take_prepared_alloc_bytes` to respond to the `native_lib` branch with `Option<MiriAllocBytes>` and then we have to do something slightly more complicated like the following again:
```rust
let alloc = alloc.adjust_from_tcx(&ecx.tcx,
    |bytes, align| ecx
        .take_prepared_alloc_bytes(id, kind, bytes, align)?
        .unwrap_or_else(|| MiriAllocBytes::from_bytes(std::borrow::Cow::Borrowed(&*bytes), align)),
    |ptr| ecx.global_root_pointer(ptr),
)?;
```

Reply:
> this call site could become a lot more simple in general:

I like that! Let's call it `get_global_alloc_bytes`.

Reply:
Nice :+1: (b3f7d2f1497bb1d2f7733407e93923dd612bd327)